Add a new behaviour to implement a fake relation with data on it

vrcAlbert opened the pull request:
an abstract model is provided to construct the fake relation

A way to use it should be like this (for an orderered relation) :

Class :

``` php
<?php
namespace Catalogue;

class Model_OrderMany extends \Nos\Model_DataMany
{
    protected static $_table_name = 'product_gamme';
    protected static $_primary_key = array('gamm__id', 'prod__id');
    protected static $_properties = array(
        'gamm__id',
        'prod__id',
        'prod_order',
    );

    protected static $_has_one = array(
        'product' => array(
            'key_from' => 'prod__id',
            'model_to' => 'Catalogue\Model_Product',
            'key_to' => 'prod_id',
            'cascade_save' => false,
            'cascade_delete' => false,
        )
    );

    protected static $_belongs_to = array(
        'gamme' => array(
            'key_from' => 'gamm__id',
            'model_to' => 'Catalogue\Model_Gamme',
            'key_to' => 'gamm_id',
            'cascade_save' => false,
            'cascade_delete' => false,
        ),
    );
}
```

Model with the behaviour : 

``` php
 protected static $_has_many = array(
        'products' => array(
            'key_from' => 'gamm_id',
            'model_to' => 'Catalogue\Model_OrderMany',
            'key_to' => 'gamm__id',
            'cascade_save' => true,
            'cascade_delete' => false,
            'conditions' => array(
                'order_by' => array('prod_order')
            )
        ),
    );

protected static $_behaviours = array(
        'Nos\Orm_Behaviour_DataMany' => array(
            'events' => array('before_save', 'after_save', 'after_delete', 'before_delete'),
            'relation' => 'products',
        ),
    );
```

Commit:
Add a new behaviour to implement a fake relation with data on it
an abstract model is provided to construct the fake relation

felixgilles commented on framework/classes/orm/behaviour/datamany.php:
```diff
@@ -0,0 +1,67 @@
+        $diff = $item->get_diff();
+        if (!empty($diff[0]) && !empty($diff[0][$relation])) {
+            $new_keys = array_keys($item->{$relation});
+            static::$_former_ids = array_diff($diff[0][$relation], $new_keys);
```
The variable static::$_former_ids is common with all instance of items that implement the behaviour. You may have multiple items that use behaviour in the same thread. 
Store the diff in a subkey by model_class::model_id

felixgilles commented on framework/classes/orm/behaviour/datamany.php:
```diff
@@ -0,0 +1,67 @@
+        $this->to_delete($item);
+    }
+
+    public function to_delete(Model $item) {
```
This method can be in camelCase

felixgilles commented on framework/classes/model/datamany.model.php:
```diff
@@ -0,0 +1,62 @@
+    public function & __get($name)
+    {
+        $relation_to = reset(static::$_has_one);
+        $relation_to_name = reset(array_keys(static::$_has_one));
```
I suggest that $relation_to_name becomes a static variable. It must be defined in the class that extends Model_DataMany

felixgilles commented on framework/classes/model/datamany.model.php:
```diff
@@ -0,0 +1,62 @@
+        $model_to = $relation_to['model_to'];
+
+        $relation_from = reset(static::$_belongs_to);
+        $relation_from_name = reset(array_keys(static::$_belongs_to));
```
Same that $relation_to_name, use a static variable. The class can have multiple belongs_to

felixgilles commented on framework/classes/model/datamany.model.php:
```diff
@@ -0,0 +1,62 @@
+
+        $relation_from = reset(static::$_belongs_to);
+        $relation_from_name = reset(array_keys(static::$_belongs_to));
+        if (empty($relation_from)) {
```
Try to make the checks in the _init() method of the class.

felixgilles commented on framework/classes/orm/behaviour/datamany.php:
```diff
@@ -0,0 +1,67 @@
+     */
+
+    public function before_save(Model $item) {
+        $relation = $this->_properties['relation'];
```
Checks that $this->_properties['relation'] is not empty in the contruct of the behaviour

felixgilles commented on framework/classes/model/datamany.model.php:
```diff
@@ -0,0 +1,62 @@
+        $model_from = $relation_from['model_to'];
+
+
+        if ((array_key_exists($name, $model_to::properties())
```
I will prefer that you call parent get() and catch OutOfBoundsException to try the same on relation_to

felixgilles commented on framework/classes/model/datamany.model.php:
```diff
@@ -0,0 +1,62 @@
+                || array_key_exists($name, $model_to::relations()))
+            && !empty($this->{$relation_to_name})) {
+            return $this->{$relation_to_name}->get($name);
+        } elseif ((array_key_exists($name, $model_from::properties())
```
Why retrieve properties on relation_from ?

Commit:
Behaviour DataMany : add class and primary key to avoid conflicts when dealing with former ids

Commit:
behaviour DataMany : camelCase in a protected method

Commit:
abstract Model DataMany : if the inherited class has several relation, it must precise the ones used to simulate the link
a good practice should be to always precise them, even if redundant

Commit:
Model DataMany : use the _init() method to set attributes or throw exception

Commit:
Behaviour DataMany : check if the property which gives the relation name exists

Commit:
if the Model_DataMany does not have a property, try to retrieve it from the related model
instead of doing the opposite

Commit:
As the ModelData model can be used for a relation having a direction, it must not be used to access the related model backward
A second model must extend the abstract model if the provided data can be use on the two different relations

eg : Order information is one which must be used in a single direction and has no sens in the other one

vrcAlbert commented:
I did as suggested but faced a problem when implementing and testing the last commit.

The idea of accessing data in one direction only is easy to understand when doing some like this : 
$categories->posts where posts is a relation on a model having a relation on a post, and a property giving the order.

In order words, doing $categories->posts allow to retrieve posts ordered in a specific category without needing to use two relations thanks to the rewrite of __get() method.

Now if I want to list all posts in a specific category thanks to the ORM and want them to be ordered by the same property, I would do something like this : 
Model_Post::query()->related('categories')->where('categories.cat_id', $id)->order_by('categories.post_order', 'ASC')->get()
which is incompatible with the idea behind the last commit.

It's a bit tricky, and I may be wrong about the way of using the ORM, please tell me what you think about this?
